### ISSUE TITLE
[bug] conan is too lenient in its specification of header_only libraries.

### ISSUE THREAD

**Issue by fulara** [conan-io/conan#6450]
~~Note: title seems to indicate that this affects only header_only but thats not the case.~~

To make matters clear I am using:
`default_package_id_mode = full_version_mode` for this example.

Consider four projects:
```
#liba:
class Conan(ConanFile):
    name = "liba"
    pass
#libb:
class Conan(ConanFile):
    name = "libb"
    requires="liba/1.0.0"

    def package_id(self):
        self.info.header_only()

#libc:
class Conan(ConanFile):
    name="libc"
    requires="libb/1.0.0"
#libd:
class Conan(ConanFile):
    name="libd"
    requires="libc/1.0.0", "liba/1.0.0"

# build initial versions of libs:
cd liba
conan create . 1.0.0@_/_
conan create . 2.0.0@_/_
cd ../libb
conan create . 1.0.0@_/_
conan create . 2.0.0@_/_
cd ../libc
conan create . 1.0.0@_/_
conan create . 2.0.0@_/_
cd ../libd
conan create . 1.0.0@_/_
```
Okay Now i have following tree built:  
`libd` build doesnt matter here because it will be our testcase.  
`libc` built with version 1.0.0 and 2.0.0 and in both depending on `libb` 1.0.0  
`libb` built with version 1.0.0 and 2.0.0 and in both depending on `liba` 1.0.0 - `libb` is header_only!  
`liba` built with version 1.0.0 and 2.0.0  

Alls good so far, now what would you expect if we overrode `liba` in `libd`?  
Well, the only things that have listed dependency on `liba` in conanfiles are `libb` and `libd`.  
Okay - but `libb` is header only - by extension - if `libb` uses any library - then that usage is baked into its `libc`.

Lets now tell `libd` to override the dependency and use `liba 2.0.0` 
```
# edit  libd' conanfile to point to liba 2.0.0
conan create . 2.0.0@_/_
```

`libd` builds without any issues. 

Conan does not require rebuild of the libc - even though it is definite that it depends on liba by proxy - author of `libc` may not even be aware that `liba` exists - but calls will be baked into its code.

What are the users expected to do in this case?
Are they to investigate all dependencies of header only libraries and copy them as is?

This is actually an extreme example of a choice I see as a problem that conan made - that dependencies are transivitely inherited, but their versioning_schema is not.  
This may make sense in some cases - for example when you dont expose in header files any types from your dependency, but its not something thats achievable and done in most of the libraries.
Conan relies on users ability to detect these, rather than providing solution. 

It is an extreme example because I am using 'header_only' here - but because of the way C++ works ( include files (... )) things that you depend on, usually automatically pollute things that depend on you.

**Comment by memsharded** [comment]
Thanks for the detailed description. I am running some tests and investigating here: https://github.com/conan-io/conan/pull/6451

**Comment by fulara** [comment]
@memsharded  I stand corrected - this seems to affect only if you use `header_only` - and that is because in `header_only` conan discards versioning_schema for its dependencies. 

In case of not `header_only` one would be required to rebuild `libb` - and then subsequently to rebuild `libc`

I was actually thinking that this is by design - thanks for fast response.


offtop:
Wow - nice test btw, if i knew that your test case would be that simple maybe I would have written above just as unit tests ~

**Comment by memsharded** [comment]
Merged #6451 this will be released in 1.23, as an opt-in in conan.conf.
